### PR TITLE
[S] Setup utils dir (#998) WIP

### PR DESCRIPTION
**Pull request by tombyers**
Still need tests to be written.

Also note comments below.

Addresses https://github.com/repositive/report-issues/issues/998

### CODE REVIEW

File: addon/mixins/r-button-mixin.js [hunk 1]
```diff
@@ -1,4 +1,5 @@
 import Ember from 'ember';
+import { getProperty } from 'ui-library/utils/set-options';
```
Comment (tombyers) [inline addon/mixins/r-button-mixin.js]: The function & the util file could be better named.

File: addon/mixins/r-button-mixin.js [hunk 2]
```diff
@@ -31,29 +32,17 @@ export default Mixin.create({
...
-   * @returns {Boolean}
-   * @private
-   */
-  _isValidFlagName(flagType, flagName) {
-    return get(this, `options.${flagType}`).indexOf(flagName) !== -1;
+    // sets appropriate properties to true for classNameBindings
+    const attributes = get(this, 'attributes');
+    const defaults = get(this, 'defaults');
+    
+    Object.keys(attributes).forEach(attribute => {
```
Comment (tombyers) [inline addon/mixins/r-button-mixin.js]: How can we reuse this logic? I wanted to keep the utils pure simple functions and not have to deal with `this` / context at all in the utility. But it meant I couldn't remove this `forEach` loop from the component.

File: addon/utils/set-options.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+/**
+* @desc returns property to set to true for classNameBinding
+* @param {Array} attributeOptions
+* @param {String} defaultOption
+* @param {String} selectedOption
+* @returns {String} property to set to true on component
+*/
+
+export function getProperty(attributeOptions, defaultOption, selectedOption) {
```
Comment (tombyers) [inline addon/utils/set-options.js]: Will do a bit of renaming today.  `option` would be better named `value`

File: addon/mixins/r-button-mixin.js [hunk 4]
```diff
@@ -31,29 +32,17 @@ export default Mixin.create({
...
+      const property = getProperty(
+        attributes[attribute],
+        defaults[attribute],
+        get(this, attribute)
+      )
+      set(this, property, true)
```
Comment (tombyers) [inline addon/mixins/r-button-mixin.js]: Two missing semi-colons - I say let's not use them at all... 😂

File: addon/mixins/r-button-mixin.js [hunk 5]
```diff
@@ -18,7 +19,7 @@ export default Mixin.create({
     'small:r-btn-small'
   ],
 
-  options: {
+  attributes: {
```
Comment (tombyers) [inline addon/mixins/r-button-mixin.js]: @SebastianPlace Does this need to be specified in init()?

Comment (SebastianPlace) [reply addon/mixins/r-button-mixin.js]: Yep, I'm working on this branch now and I've already changed this 😄

File: addon/components/r-avatar/component.js [hunk 6]
```diff
@@ -21,24 +21,17 @@ export default Component.extend({
...
+    this.allowedAttributes = {
       size: ['small', 'medium', 'large', 'x-large']
     };
     this.defaults = { size: 'medium' };
 
-    // sets appropriate properties to true for classNameBindings
-    const attributes = get(this, 'attributes');
+    const allowedAttributes = get(this, 'allowedAttributes');
     const defaults = get(this, 'defaults');
+    const suppliedAttrs = get(this, 'attrs');
```
Comment (tombyers) [inline addon/components/r-avatar/component.js]: What is attrs? Are you proposing changing the component interface? e.g
```
{{r-button 
  attrs={
    size: 'small',
    variant: 'secondary'
  }
}}
```

Comment (SebastianPlace) [reply addon/components/r-avatar/component.js]: No. Attrs is the object on the component where the parameters passed into the components are held. So specifying `{{r-button label='test' size='big' variant='primary'}}` would mean `attrs`= 
```
{
  label: 'test',
  size: 'big',
  variant: 'primary'
}
```

File: tests/unit/utils/create-attributes-object-test.js [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+      suppliedAttrs
+    )).to.have.all.keys('secondary', 'medium', 'green');
+  });
+
+  it('should filter out invalid attributes', function() {
+    const suppliedAttrs = {
+      color: 'red',
+      fruit: "banana",
+      veg: 'onion',
+      cereal: "Cheerios"
```
Comment (SebastianPlace) [inline tests/unit/utils/create-attributes-object-test.js:64]: 👌

**Review by SebastianPlace — APPROVED** [review]
@tombyers great work on the tests!

Everything looks good to me 👍

**Comment by SebastianPlace** [comment]
@tombyers Go ahead and merge this when you're happy

**Comment by tombyers** [comment]
Whoops, clicked merge without using our naming convention. Sorry for making our changelog ugly!

**Comment by Istar-Eldritch** [comment]
An admin should be able to override it. We have to discuss about it at some point

**Pull request merged**